### ISSUE TITLE
Bolded and Italics code does not render properly on mobile 

### ISSUE THREAD

**Issue by NicMendonca** [Expensify/Expensify.cash#2491]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
Bold and italics are applied to the code block

## Actual Result:
Bold and italics are not applied to the code block

## Action Performed:
1. Send the following message on mobile
```
*`test`*
_`test`_
```
2. See how italics and bold are not applied to the code section
3. View the message on desktop / web
4. See how italics and bold are applied 

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platform:
Where is this issue occurring?

Mobile, reproduced on both - 
 iOS
 Android

**Version Number:**  1.0.2-74
**Notes/Photos/Videos:** 

Bold / Strikethrough on mobile
![image](https://user-images.githubusercontent.com/22447860/113033388-d1966f00-9145-11eb-9dea-ae2a3527edf9.png)

Bold/ Strikethrough on desktop
![image](https://user-images.githubusercontent.com/22447860/113033424-dbb86d80-9145-11eb-8ef9-e6db8d7aa94a.png)

Video: https://recordit.co/QtKls2EXDi

**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/158825

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

Upwork job: https://www.upwork.com/jobs/~016d9851436c63f135

**Comment by parasharrajat** [comment]
@NicMendonca This is intended behavior and defined in the code.

**Comment by aliabbasmalik8** [comment]
@NicMendonca We can fix this by updating `RenderHTML` component like this. tested on IOS and WEB and it's working accordingly 

![image](https://user-images.githubusercontent.com/31027036/115488070-e7202580-a227-11eb-89fe-f421e43e663c.png)

and here is classesStyles that placed in `webViewStyles` object.
```
classesStyles: {
        em: {
            fontFamily: fontFamily.GTA_ITALIC,
            fontStyle: italic,
        },

        strong: {
            fontFamily: fontFamily.GTA_BOLD,
            fontWeight: fontWeightBold,
        },

    },
```

**Comment by thierrysantos** [comment]
@NicMendonca

I spend a time debugging here and I found that in android we are not using the correct values.

When text must be italic, it is changed to `normal`.
When text must be 700 (bold), it is changed `600` (semibold).

That is why it is not showing correctly, you could see it below.

https://github.com/Expensify/Expensify.cash/blob/9efcc63e4eb9dd27fac35d7bfc92b3400554be0c/src/styles/fontWeight/bold/index.android.js#L3
https://github.com/Expensify/Expensify.cash/blob/9efcc63e4eb9dd27fac35d7bfc92b3400554be0c/src/styles/italic/index.android.js#L1

In the begging of the file, the @marcaaron explain why

> Android has GTAmericaExp-Bold, but fontWeight: '700' will result in an incorrect font displaying on Android.

I changed to use bold and italic.

! [Screen Shot 2021-04-21 at 14 12 36] (https://user-images.githubusercontent.com/14885967/115594013-a10ca580-a2ab-11eb-8ffe-ccae828d23f1.png)

**Comment by parasharrajat** [comment]
@thierrysantos LOL. This issue literally killed me. I graved out the `react-native-render-html`'s code. What a absolutely waste of time.
Why `italic = normal`?????

Nice found. Great work.  :tada:

**Comment by marcaaron** [comment]
In the screenshots the type is no longer monospaced ?

**Comment by thierrysantos** [comment]
@marcaaron yeah, when using bold or italic the text isn't monospaced

**Comment by parasharrajat** [comment]
So we should load a monospaced italic font for fontstyle italic.

**Comment by marcaaron** [comment]
I think we want them to be the same on all platforms cc @shawnborton have you got the right font to use on hand ?

**Comment by parasharrajat** [comment]
GTAmericaExpMono-Rg.otf should have the italic and bold variant.

**Comment by thierrysantos** [comment]
> @thierrysantos LOL. This issue literally killed me. I graved out the react-native-render-html's code. What a absolutely waste of time.

Yes, if I didn't find that, looking at the react-native-render-html code would be the next step 😂

**Comment by thierrysantos** [comment]
> GTAmericaExpMono-Rg.otf should have the italic and bold variant.

I try with that and have the same result 🤔

**Comment by parasharrajat** [comment]
I mean it should have the italic and bold-variant on the internet which we can use something line GTAmericaExpMono-Italic.otf.

**Comment by thierrysantos** [comment]
Yeah, I found some variants here https://www.grillitype.com/typeface/gt-america

**Comment by parasharrajat** [comment]
@shawnborton would be a good person to get the assets.

**Comment by thierrysantos** [comment]
nice, I'm waiting to the next steps 🤘🏽

**Comment by aliabbasmalik8** [comment]
Hi @thierrysantos 
this solution did not work on IOS for me. let me know if it's working for you. Thanks
https://github.com/Expensify/Expensify.cash/issues/2491#issuecomment-824224404

**Comment by Luke9389** [comment]
Hello @thierrysantos! I'll be reviewing your proposal on this issue. The description of the issue does include both Android and iOS, and I see that you've cracked part of the problem on Android. Can you account for why this is happening on iOS, and what you'd change to fix it?

**Comment by Luke9389** [comment]
@aliabbasmalik8 I'll also review your proposal. Can you be more specific about the exact file you'll be changing? Also I'm not sure that this addresses the root cause of the issue. It should not be necessary to apply styles in this case, considering that the following HTML all works fine. (you can try it here: https://www.w3schools.com/html/tryit.asp?filename=tryhtml_intro)
```
<strong><code>Hello</code></strong><br/>
<em><code>Hello</code><em><br>
<code>Hello</code><br/>
Hello
```

**Comment by aliabbasmalik8** [comment]
@Luke9389 
these changes made in these files. `RenderHTML.js` && `src/styles/styles.js` and classesStyles  added in `webViewStyles` style object
https://github.com/Expensify/Expensify.cash/issues/2491#issuecomment-823728497

**Comment by Luke9389** [comment]
@aliabbasmalik8 
Ok but that really just addresses my surface level concern about the level of detail in the proposal. I'm still not convinced that it's a good idea to add styles in this case. It seems like we're fixing this downstream of the actual problem.

**Comment by aliabbasmalik8** [comment]
@Luke9389 
`<strong>`, `<code>`, `<em>` etc,  these tags achieved in react native by adding styling in the `<Text>` component like you can see in the code. so when we have code like this `<strong><code>` then inner tag style applying so I updated inner tag styling.
![image](https://user-images.githubusercontent.com/31027036/115609799-ff3e8600-a2b5-11eb-97ec-05c1cd71c9ac.png)

**Comment by Luke9389** [comment]
@aliabbasmalik8
Ah I see. So you're saying that the fontFamily style is being over written because we are applying it in two places (e.g. `code` and `em`)?

**Comment by aliabbasmalik8** [comment]
> @aliabbasmalik8
> Ah I see. So you're saying that the fontFamily style is being over written because we are applying it in two places (e.g. `code` and `em`)?

yes

**Comment by Luke9389** [comment]
@aliabbasmalik8 
Ah, ok gotcha. Thanks for helping me understand here. 
Have you tested this on Android? I don't think it'll work because Android doesn't like `fontStyle: italic `. 
Here is a link to the original change that @thierrysantos found above: https://github.com/Expensify/Expensify.cash/pull/549#discussion_r498356202

Seems like the fontStyle will need to remain platform dependent, but your change will be overwriting that. Correct?
Maybe it'd be good for @sketchydroide to weigh in here.

**Comment by thierrysantos** [comment]
I'm thinking here and probably the problem is related to font variant.

![Screen Shot 2021-04-21 at 16 56 29](https://user-images.githubusercontent.com/14885967/115613249-8c87d780-a2c2-11eb-8197-c5619475798f.png) -> Without font monospace

When we are using monospace(`regular`), when the text is rendered as bold it doesn't occurs.

![Screen Shot 2021-04-21 at 16 55 08](https://user-images.githubusercontent.com/14885967/115613446-c5c04780-a2c2-11eb-9ee0-16f03bd63401.png)

**Comment by aliabbasmalik8** [comment]
@Luke9389 
for android, we already have a separate file as you can see `italic/index.android.js ` 
```
const italic = 'normal';
export default italic;
```
but I think we also need to add `fontFamily.MONOSPACE` `Italic` and `bold` version. right now I test with these styling and fonts family
```
classesStyles: {
        em: {
            fontFamily: fontFamily.GTA_ITALIC,
            fontStyle: italic,
        },

        strong: {
            fontFamily: fontFamily.GTA_BOLD,
            fontWeight: fontWeightBold,
        },

    },
```

**Comment by thierrysantos** [comment]
Also we could see here some issues related to styling https://github.com/Expensify/Expensify.cash/issues/2491#issuecomment-824322482

**Comment by shawnborton** [comment]
Just to make sure I understand this correctly, it sounds like we need our monospaced font in regular italic, bold, and bold italic. Does that sound right? We can provide the font files to implement these natively, and I think we'll also have to make sure these fonts are added to our Expensify.com web application so we can serve them up for desktop/web.

**Comment by Luke9389** [comment]
@shawnborton Yes I think that's correct. 

@aliabbasmalik8 I'm liking the direction you're heading, but you'll need to _verify_ that your proposed solution works for iOS AND Android (with screenshots). If you can do that, please resubmit a proposal with those included, and a summary of exactly what you're going to change, links to those files, and any resources you'll need (like a new font for example). Please use links to our actual codebase ([example](https://github.com/Expensify/Expensify.cash/issues/2491#issuecomment-824224404)), and markdown code blocks rather than screenshots of your IDE. Generally speaking, I'm looking for a very clean proposal that has everything we need, all in one comment. 

@thierrysantos I'm not quite following your last two comments, but I ask that you please only comment when you have a solution ready and have a proposal to resubmit. Posting investigation steps is making this thread noisy and hard to follow.

**Comment by thierrysantos** [comment]
Sorry, if I'm not being clear enough, but what I'm suggesting here (https://github.com/Expensify/Expensify.cash/issues/2491#issuecomment-824322482 and https://github.com/Expensify/ Expensify .cash / issues / 2491 # issuecomment-824237403) is to have other variants for monospace font, such as bold and italics @Luke9389

**Comment by shawnborton** [comment]
Here are the font files we need for native implementation: [GTAMono.zip](https://github.com/Expensify/Expensify.cash/files/6354227/GTAMono.zip)

As I mentioned, we'll need to implement these on Web-E as well as I am pretty sure desktop/web on Expensfiy.cash just grabs font URLs from the www.expensify.com web app. @Luke9389 is that something you would like to take care of? I can help you implement them as well.

**Comment by aliabbasmalik8** [comment]
@Luke9389 
we need to add GTAmericaExpMono font in regular italic, bold, and bold italic.
and need to delete `fontWeight/bold/index.android.js` && `fontWeight/italic/index.android.js`

NOTE: 

1. I tested with `OpenSans` free fonts instead `GTAmericaExpMono` fonts and it worked for me.
2. below images are used `OpenSans` free fonts instead `GTAmericaExpMono`  fonts for testing purpose

IOS
<img width="396" alt="Screen Shot 2021-04-21 at 6 12 44 PM" src="https://user-images.githubusercontent.com/31027036/115627552-39675200-a2cd-11eb-885d-87ea13d7c50d.png">

**Comment by aliabbasmalik8** [comment]
@Luke9389 
screenshots with provided fonts
IOS

![image](https://user-images.githubusercontent.com/31027036/115628284-6d8f4280-a2ce-11eb-807c-0a81297bde0a.png)

Android

![image](https://user-images.githubusercontent.com/31027036/115628596-d971ab00-a2ce-11eb-8f57-cccfa31409fd.png)

**Comment by shawnborton** [comment]
@aliabbasmalik8 for your testing, can you be sure to use words with a lowercase "g" in them? That makes it easy to tell if the correct font is being applied. 

For Android in your screenshot above, the fonts look like they are incorrect.

**Comment by aliabbasmalik8** [comment]
> @aliabbasmalik8 for your testing, can you be sure to use words with a lowercase "g" in them? That makes it easy to tell if the correct font is being applied.
> 
> For Android in your screenshot above, the fonts look like they are incorrect.

Okay,

**Comment by aliabbasmalik8** [comment]
@shawnborton  You are right. fonts did not add in android. I have no idea why it's happened but i linked all assets in app by running `npx react-native link`.

**Comment by thierrysantos** [comment]
Hello @Luke9389 ,

I have a proposal right now, after a lot of time debugging we need to add this new fonts and rename it based on `PostScript name`.

We need to add this fonts here

https://github.com/Expensify/Expensify.cash/blob/d27cf29a5579cbaab20fd4c29ad92140f55580e8/src/styles/fontFamily/index.js#L3

and here

https://github.com/Expensify/Expensify.cash/blob/d27cf29a5579cbaab20fd4c29ad92140f55580e8/src/components/RenderHTML.js#L20

Also we need to change the `strong` and `em` to be:

```
        em: {
            fontFamily: fontFamily.GTA_ITALIC,
            fontStyle: 'italic',
        },
        
         strong: {
            fontFamily: fontFamily.GTA_BOLD,
            fontWeight: 'bold',
        },
```

And in the <CodeRenderer /> we need to load the font variant dynamically (only for android), because react-native in android don't work well with fontWeight and fontStyle.

https://github.com/Expensify/Expensify.cash/blob/d27cf29a5579cbaab20fd4c29ad92140f55580e8/src/components/RenderHTML.js#L62

The code below is not the end, and it is just to show the solution, I will improve it

```
    if (Platform.OS === 'android') {
        const isItalic = textStyle.fontStyle === 'italic';
        const isBold = textStyle.fontWeight === 'bold' || Number(textStyle.fontWeight) >= 700;
        const isItalicAndBold = isItalic && isBold;

        const fontName = isItalicAndBold
            ? fontFamily.MONOSPACE_BOLD_ITALIC
            : isItalic
                ? fontFamily.MONOSPACE_ITALIC
                : isBold ? fontFamily.MONOSPACE_BOLD : fontFamily.MONOSPACE;
        customStyle = {
            ...customStyle,
            fontFamily: fontName,
            fontStyle: undefined, // To don't use the element fontStyle
            fontWeight: undefined, // To don't use the element fontWeight
        };
    }
```

That is the final result:

<img width="981" alt="Screen Shot 2021-04-23 at 01 58 59" src="https://user-images.githubusercontent.com/14885967/115822200-85ec8380-a3da-11eb-9208-4d571eb31307.png">

**Comment by aliabbasmalik8** [comment]
**REVISED PROPOSAL**
**STEP 1:** ADD FONTS IN APP AND LINKED
**STEP 2:** ADD FONT FAMILY CONSTANTS AT HERE
https://github.com/Expensify/Expensify.cash/blob/d27cf29a5579cbaab20fd4c29ad92140f55580e8/src/styles/fontFamily/index.js#L3
```
    MONOSPACE_BOLD: 'GTAmericaExpMono-Bold',
    MONOSPACE_ITALIC: 'GTAmericaExpMono-RgIt',
    MONOSPACE_BOLD_ITALIC: 'GTAmericaExpMono-BdIt',
```
**STEP 4**: DELETE THESE 2 files
- https://github.com/Expensify/Expensify.cash/blob/main/src/styles/fontWeight/bold/index.android.js
- https://github.com/Expensify/Expensify.cash/blob/main/src/styles/italic/index.android.js

**STEP 5**: ADD CODE HERE
https://github.com/Expensify/Expensify.cash/blob/d27cf29a5579cbaab20fd4c29ad92140f55580e8/src/components/RenderHTML.js#L156
```
    // added classed to add appropraite fontfamily
    let updatedHTML = html.replace(/<em><code>/g, '<em><code class="em">');
    updatedHTML = updatedHTML.replace(/<strong><code>/g, '<strong><code class="strong">');
    updatedHTML = updatedHTML.replace(/<code>/g, '<code class="code">');
 ```
 
 **STEP 6:** ADD `classesStyles` PROPS IN `HTML` COMPONENT
https://github.com/Expensify/Expensify.cash/blob/d27cf29a5579cbaab20fd4c29ad92140f55580e8/src/components/RenderHTML.js#L158
 ```
             html={updatedHTML}
            classesStyles={webViewStyles.classesStyles}
 ```
 **STEP 7:** UPDATE `styles.js`
- REMOVE `fontFamily: fontFamily.MONOSPACE,` from `code`object at this point
https://github.com/Expensify/Expensify.cash/blob/d27cf29a5579cbaab20fd4c29ad92140f55580e8/src/styles/styles.js#L1424
- ADD `classesStyles` object `webViewStyles` styles object
https://github.com/Expensify/Expensify.cash/blob/d27cf29a5579cbaab20fd4c29ad92140f55580e8/src/styles/styles.js#L1372
```
classesStyles: {
        em: {
            fontFamily: fontFamily.MONOSPACE_ITALIC,
        },
        strong: {
            fontFamily: fontFamily.MONOSPACE_BOLD,
        },
        code: {
            fontFamily: fontFamily.MONOSPACE,
        },
    },
```

**SCREENSHOTS**
**IOS**
![image](https://user-images.githubusercontent.com/31027036/115857533-d0322c80-a3fb-11eb-8a24-0a02120f0944.png)
**ANDROID**
![IMG_20210423_152954](https://user-images.githubusercontent.com/31027036/115858855-76326680-a3fd-11eb-87d7-5bad62802094.jpg)

**Comment by Luke9389** [comment]
@thierrysantos I see you've got platform-specific code in your proposal. Could you describe to me at a high level how you'd improve your code?

**Comment by thierrysantos** [comment]
@Luke9389 that idea is just refactor the code to remove the nested ternary. Like this

```
const italic = textStyle.fontStyle === 'italic' && fontFamily.MONOSPACE_ITALIC
const bold = textStyle.fontWeight === 'bold'&& fontFamily.MONOSPACE_BOLD
const italicBold = italic && bold && fontFamily.MONOSPACE_BOLD_ITALIC

const fontFamily = italicBold || bold || italic || fontFamily.MONOSPACE
```

**Comment by Luke9389** [comment]
@thierrysantos I see. But what about `Platform.OS === 'android'`? In our contributing guidelines, we mention that we don't want platform-specific code. Is there a way to avoid having that line? Perhaps we could put this logic in a lib and import it to `RenderHTML` instead?

**Comment by thierrysantos** [comment]
@Luke9389 I'm looking a way here and I give a feedback for you

**Comment by thierrysantos** [comment]
@Luke9389 we don't need this `Platform.OS === 'android'`, so is just that

```
const italic = textStyle.fontStyle === 'italic' && fontFamily.MONOSPACE_ITALIC
const bold = textStyle.fontWeight === 'bold'&& fontFamily.MONOSPACE_BOLD
const italicBold = italic && bold && fontFamily.MONOSPACE_BOLD_ITALIC

const fontFamily = italicBold || bold || italic || fontFamily.MONOSPACE
```

**Comment by Luke9389** [comment]
Ok, @thierrysantos you can go ahead and start the PR. Thanks for all your patience here.

@aliabbasmalik8 I appreciate the effort you put into this thread, but I've decided not to move forward with your proposal. The main reason for this is that after I asked for a clear, complete, and well laid out proposal from both of you, Thierry was the one to provide that first. I do hope that you continue to make proposals with us, but in the future, I think you should focus less on speed, and more on making sure your proposal is clear and complete. The revised version you did was good, had you posted that much sooner this would have gone the other way. Thank you again very much for your time and efforts.

**Comment by thierrysantos** [comment]
> Ok, @thierrysantos you can go ahead and start the PR. Thanks for all your patience here.

Okay, thanks and sorry if I made something wrong here ❤️ .

**Comment by Luke9389** [comment]
No, you've been great! No worries at all.
The process needs all of us :)

**Comment by thierrysantos** [comment]
Nice, I sent the proposal on UpWork

**Comment by NicMendonca** [comment]
@thierrysantos just sent over the the offer in Upwork. Thanks!

**Comment by thierrysantos** [comment]
@Luke9389 The PR is open 🎉  https://github.com/Expensify/Expensify.cash/pull/2648
